File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        int maxlen=0;
        int l=0,r=0;
        unordered_map<char,int>mpp;
        while(r<n){
            if(mpp.find(s[r])!=mpp.end()){
                if(mpp[s[r]]>=l){
                l=mpp[s[r]]+1;
                }
            }
            int len=r-l+1;
            mpp[s[r]]=r;
            maxlen=max(maxlen,len);
            r++;
            
        }
        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int left = 0;
        int maxi = 0;
        Map<Character, Integer> mpp = new HashMap<>();
        for(int i =0;i<n;i++){
            if(mpp.containsKey(s.charAt(i))){
                while(s.charAt(left)!=s.charAt(i)){
                        mpp.remove(s.charAt(left));
                       left++;
                }
                mpp.remove(s.charAt(left));
                left++;
            }
            mpp.put(s.charAt(i), mpp.getOrDefault(s.charAt(i), 0) + 1);
            maxi = Math.max(maxi,i-left+1);
        }
        return maxi;
    }
}